### PR TITLE
Add line # of previous definition (resubmit #3396)

### PR DESCRIPTION
**Pull request by pkch**
Partial fix for #1874: improve the error message. The duplicate definitions in different if/else branches are still not allowed (this is much harder to fix).

This is a resubmission of #3396 which caused the crash in #3415 and was reverted. The old PR is in commit 39c3058; after that, I added 2 more commits: the failing test that repros the crash, and the fix to it.

### CODE REVIEW

**Comment by pkch** [comment]
In case the context is `None`, or line number is -1, I put a message suggesting that the user look at  imports. Is it ok? There could be other cases where this branch is hit that I don't list, but I thought this partial suggestion might still help.

File: mypy/semanal.py [hunk 1]
```diff
@@ -3324,8 +3324,16 @@ def name_not_defined(self, name: str, ctx: Context) -> None:
...
 
-    def name_already_defined(self, name: str, ctx: Context) -> None:
-        self.fail("Name '{}' already defined".format(name), ctx)
+    def name_already_defined(self, name: str, ctx: Context,
+                             original_ctx: Optional[SymbolTableNode] = None) -> None:
+        if original_ctx:
+            if original_ctx.node and original_ctx.node.get_line() != -1:
+                extra_msg = ' on line {}'.format(original_ctx.node.get_line())
+            else:
+                extra_msg = ' (line unavailable; possibly an import)'
```
Comment (JukkaL) [inline mypy/semanal.py]: We could simplify the message a bit. What about just ` (possibly by an import)`. The 'line unavailable' part doesn't provide useful information really.

File: test-data/unit/semanal-errors.test [hunk 2]
```diff
@@ -1330,3 +1330,13 @@ a = ('spam', 'spam', 'eggs', 'spam')  # type: Tuple[str]
...
+
+[case testUnknownContextDuplicateDef]
+from m import A
+class A:
+    pass
+[file m.py]
+class A:
+    pass
+[out]
+main:2: error: Name 'A' already defined (line unavailable; possibly an import)
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test]: Here are further ideas for test cases:

* Original definition is a decorated function.
* Original definition is an overloaded function.
* Original definition is a named tuple definition.
* Original definition is a typed dict definition.
* Original definition is a module (`import m`).

Comment (pkch) [inline test-data/unit/semanal-errors.test]: If a module is simply imported twice, there's no error message. Did you mean some other test?

Comment (ilevkivskyi) [inline test-data/unit/semanal-errors.test]: @pkch Maybe Jukka means something like this:
```python
import re
re = 1
```

Comment (pkch) [inline test-data/unit/semanal-errors.test]: Yes, but that will cause a completely different error message (types are not compatible). It won't get to the point of complaining about duplicate definition.

Comment (ilevkivskyi) [inline test-data/unit/semanal-errors.test]: And what about
```python
import re
import math
re = math
```

Comment (pkch) [inline test-data/unit/semanal-errors.test]: In this case, there's no error message at all (as expected?).

File: test-data/unit/semanal-errors.test [hunk 3]
```diff
@@ -1330,3 +1330,68 @@ a = ('spam', 'spam', 'eggs', 'spam')  # type: Tuple[str]
...
+
+[case testDuplicateDefFromImport]
+from m import A
+class A:
+    pass
+[file m.py]
+class A:
+    pass
+[out]
+main:2: error: Name 'A' already defined (possibly by an import)
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test]: Can you use `# E: ...` comments in these cases instead?

File: test-data/unit/semanal-errors.test [hunk 4]
```diff
@@ -1330,3 +1330,68 @@ a = ('spam', 'spam', 'eggs', 'spam')  # type: Tuple[str]
...
+from mypy_extensions import TypedDict
+Point = TypedDict('Point', {'x': int, 'y': int})
+
+class Point:
+    pass
+[builtins fixtures/dict.pyi]
+
+[out]
+main:4: error: Name 'Point' already defined on line 2
+
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test]: Remove extra empty lines at the end of file.

File: test-data/unit/semanal-errors.test [hunk 5]
```diff
@@ -1330,3 +1330,68 @@ a = ('spam', 'spam', 'eggs', 'spam')  # type: Tuple[str]
...
+[case testDuplicateDefTypedDict]
+from mypy_extensions import TypedDict
+Point = TypedDict('Point', {'x': int, 'y': int})
+
+class Point:
+    pass
+[builtins fixtures/dict.pyi]
+
+[out]
+main:4: error: Name 'Point' already defined on line 2
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test]: Suggestions for additional test cases:
* `import m` followed by `def m(): ...`
* `T = TypeVar('T')` followed by `class T: ...`
* Ignored import (`import m # type: ignore` or `from m import f # type: ignore`) followed by definition.
* Type alias like `A = List[int]` followed by a redefinition.

File: test-data/unit/semanal-errors.test [hunk 6]
```diff
@@ -1330,3 +1330,68 @@ a = ('spam', 'spam', 'eggs', 'spam')  # type: Tuple[str]
...
+    def f(x: str) -> None:
+        pass
+    def f(x: Any) -> None:
+        pass
+else:
+    def f(x: str) -> None:
+        pass
+[out]
+main:12: error: Name 'f' already defined on line 3
+
```
Comment (JukkaL) [inline test-data/unit/semanal-errors.test]: Remove second empty line (here and below).

**Comment by pkch** [comment]
This may be ready to merge (the last CR fixes was only to add more tests, no code changes).

**Pull request merged**